Style engine: enqueue block support styles

ramonjd opened the pull request:
## What?

This PR extends the backend style engine, and:

- enables the storing and enqueuing of layout and elements block support styles.
- batches layout and elements styles on the backend into a single HTML style tag 

## Why?
This is part of the effort to reduce the number of style tags in the rendered HTML. 

## How?
Using the classes laid down by @aristath in https://github.com/WordPress/gutenberg/pull/42222, it's now possible to build stylesheets and enqueue CSS for rendering on the frontend.

There are three global functions that help to achieve this:

### wp_style_engine_get_styles()

Use this function for style objects that require parsing, for instance, the value of a block's attributes.style object or the top level styles in theme.json.

To enqueue a style for rendering in the frontend, the `$options` array requires the following:

1. **selector (string)** - this is the CSS selector for your block style CSS declarations.
2. **context (string)** - this tells the style engine where to store the styles. Styles in the same context will be batched together and printed in the same HTML style tag. The default is `'block-supports'`.
3. **enqueue (boolean)** - tells the style engine to enqueue the styles and print them in the frontend.

`wp_style_engine_get_styles` will return the compiled CSS and CSS declarations array.

```php
$block_styles =  array(
     'spacing' => array( 'padding' => '100px' )
);
$styles = wp_style_engine_get_styles(
    $block_styles,
    array(
        'selector' => '.a-selector',
        'context'  => 'block-supports',
        'enqueue'  => true,
    )
);
print_r( $styles );

/*
array(
    'css'                 => '.a-selector{padding:10px}'
    'declarations'  => array( 'padding' => '100px' )
)
*/
```

### wp_style_engine_add_to_store()

Use this function to enqueue any CSS rules. It will automatically merge declarations and combine selectors.

```php
$styles = array(
    array(
        'selector'.       => '.wp-pumpkin',
        'declarations' => array( 'color' => 'orange' )
    ),
    array(
        'selector'.       => '.wp-tomato',
        'declarations' => array( 'color' => 'red' )
    ),
    array(
        'selector'.       => '.wp-tomato',
        'declarations' => array( 'padding' => '100px' )
    ),
    array(
        'selector'.       => '.wp-kumquat',
        'declarations' => array( 'color' => 'orange' )
    ),
);

wp_style_engine_add_to_store( 'layout-block-supports', $styles );


```

The resulting stylesheet will be:
```html
<style id='layout-block-supports-inline-css'>
.wp-pumpkin, .wp-kumquat {color:orange}.wp-tomato{color:red;padding:100px}
</style>
```

### wp_style_engine_get_stylesheet_from_css_rules()

Use this function to compile and return a stylesheet for any CSS rules. This function does not enqueue styles, but rather acts as a CSS compiler.

It has a similar function signature to `wp_style_engine_add_to_store`,

```php
$styles = array(
    array(
        'selector'.       => '.wp-pumpkin',
        'declarations' => array( 'color' => 'orange' )
    ),
    array(
        'selector'.       => '.wp-tomato',
        'declarations' => array( 'color' => 'red' )
    ),
    array(
        'selector'.       => '.wp-tomato',
        'declarations' => array( 'padding' => '100px' )
    ),
    array(
        'selector'.       => '.wp-kumquat',
        'declarations' => array( 'color' => 'orange' )
    ),
);

$stylesheet = wp_style_engine_get_stylesheet_from_css_rules( 'layout-block-supports', $styles );
print_r( $stylesheet ); // .wp-pumpkin, .wp-kumquat {color:orange}.wp-tomato{color:red;padding:100px}
```

## Future iterations

- Move enqueue logic into Gutenberg
- Add an optimize (and possibly a prettify) flag to toggle selector combination optimization and whitespace
  
## Testing Instructions

This PR enqueues both link elements styles and Layout block support styles.

In a new post add some text links and assign them a color. 

Insert a few group blocks and adjust the layout settings (alignment, widths, justifications). Ideally, manually test each of the features of both default/flow and flex-based Layouts using blocks such as Group, Columns, Buttons. 

Ensure that the colors and layout appear on the frontend as expected.

Here's some test block HTML. I'm testing using emptytheme, but we should also test in other themes including classic themes.

<details>

<summary>Example markup for custom link colors</summary>

```html
<!-- wp:paragraph -->
<p><a href="https://test.com">Test</a> paragraph with a link.</p>
<!-- /wp:paragraph -->

<!-- wp:paragraph {"style":{"elements":{"link":{"color":{"text":"#842ecf"}}}}} -->
<p class="has-link-color"><a href="https://test.com">Test</a> paragraph with a link.</p>
<!-- /wp:paragraph -->

<!-- wp:paragraph {"style":{"elements":{"link":{"color":{"text":"#842ecf"}}}}} -->
<p class="has-link-color"><a href="https://test.com">Test</a> paragraph with a link.</p>
<!-- /wp:paragraph -->

<!-- wp:paragraph {"style":{"elements":{"link":{"color":{"text":"var:preset|color|pale-pink"}}}}} -->
<p class="has-link-color"><a href="https://test.com">Test</a> paragraph with a link.</p>
<!-- /wp:paragraph -->

<!-- wp:paragraph {"style":{"elements":{"link":{"color":{"text":"var:preset|color|pale-pink"}}}}} -->
<p class="has-link-color"><a href="https://test.com">Test</a> paragraph with a link.</p>
<!-- /wp:paragraph -->

<!-- wp:heading {"style":{"elements":{"link":{"color":{"text":"var:preset|color|pale-pink"}}}}} -->
<h2 class="has-link-color">H2 with a <a href="https://wordpress.org">link</a></h2>
<!-- /wp:heading -->

<!-- wp:group {"style":{"elements":{"link":{"color":{"text":"#fbc941"}}}}} -->
<div class="wp-block-group has-link-color"><!-- wp:paragraph -->
<p><a href="https://wordpress.org">Test</a> paragraph with a link inside a group;</p>
<!-- /wp:paragraph --></div>
<!-- /wp:group -->

<!-- wp:group {"style":{"elements":{"link":{"color":{"text":"#fbc941"}}}}} -->
<div class="wp-block-group has-link-color"><!-- wp:paragraph -->
<p><a href="https://wordpress.org">Test</a> paragraph with a link inside a group;</p>
<!-- /wp:paragraph --></div>
<!-- /wp:group -->

<!-- wp:group {"backgroundColor":"luminous-vivid-amber","layout":{"inherit":true}} -->
<div class="wp-block-group has-luminous-vivid-amber-background-color has-background"><!-- wp:columns -->
<div class="wp-block-columns"><!-- wp:column {"width":"47px","layout":{"contentSize":"26px"}} -->
<div class="wp-block-column" style="flex-basis:47px"><!-- wp:paragraph -->
<p>testest</p>
<!-- /wp:paragraph --></div>
<!-- /wp:column -->

<!-- wp:column -->
<div class="wp-block-column"><!-- wp:paragraph -->
<p>test</p>
<!-- /wp:paragraph --></div>
<!-- /wp:column --></div>
<!-- /wp:columns --></div>
<!-- /wp:group -->

<!-- wp:group {"backgroundColor":"luminous-vivid-orange","layout":{"inherit":false,"contentSize":"402px"}} -->
<div class="wp-block-group has-luminous-vivid-orange-background-color has-background"><!-- wp:columns -->
<div class="wp-block-columns"><!-- wp:column {"width":"47px","layout":{"contentSize":"26px"}} -->
<div class="wp-block-column" style="flex-basis:47px"><!-- wp:paragraph -->
<p>test</p>
<!-- /wp:paragraph --></div>
<!-- /wp:column -->

<!-- wp:column -->
<div class="wp-block-column"><!-- wp:paragraph -->
<p>test</p>
<!-- /wp:paragraph --></div>
<!-- /wp:column --></div>
<!-- /wp:columns --></div>
<!-- /wp:group -->

<!-- wp:group {"backgroundColor":"luminous-vivid-orange","layout":{"type":"flex","flexWrap":"nowrap"}} -->
<div class="wp-block-group has-luminous-vivid-orange-background-color has-background"><!-- wp:columns -->
<div class="wp-block-columns"><!-- wp:column {"width":"47px","layout":{"contentSize":"26px"}} -->
<div class="wp-block-column" style="flex-basis:47px"><!-- wp:paragraph -->
<p>test</p>
<!-- /wp:paragraph --></div>
<!-- /wp:column -->

<!-- wp:column -->
<div class="wp-block-column"><!-- wp:paragraph -->
<p>test</p>
<!-- /wp:paragraph --></div>
<!-- /wp:column --></div>
<!-- /wp:columns --></div>
<!-- /wp:group -->

<!-- wp:group {"backgroundColor":"vivid-purple"} -->
<div class="wp-block-group has-vivid-purple-background-color has-background"><!-- wp:paragraph -->
<p>test</p>
<!-- /wp:paragraph --></div>
<!-- /wp:group -->
```

</details>

<details>

<summary>Example markup for Layout support</summary>

```html
<!-- wp:group {"align":"full","style":{"spacing":{"padding":{"top":"30px","right":"30px","bottom":"30px","left":"30px"}}},"backgroundColor":"vivid-red","layout":{"contentSize":"600px"}} -->
<div class="wp-block-group alignfull has-vivid-red-background-color has-background" style="padding-top:30px;padding-right:30px;padding-bottom:30px;padding-left:30px"><!-- wp:cover {"overlayColor":"primary","align":"full"} -->
<div class="wp-block-cover alignfull"><span aria-hidden="true" class="wp-block-cover__background has-primary-background-color has-background-dim-100 has-background-dim"></span><div class="wp-block-cover__inner-container"><!-- wp:paragraph {"align":"center","placeholder":"Write title…","fontSize":"large"} -->
<p class="has-text-align-center has-large-font-size">A cover block that's full width to the edge</p>
<!-- /wp:paragraph --></div></div>
<!-- /wp:cover --></div>
<!-- /wp:group -->

<!-- wp:buttons {"layout":{"type":"flex","flexWrap":"nowrap"},"style":{"spacing":{"blockGap":"50px"}}} -->
<div class="wp-block-buttons"><!-- wp:button -->
<div class="wp-block-button"><a class="wp-block-button__link wp-element-button">Button</a></div>
<!-- /wp:button -->

<!-- wp:button -->
<div class="wp-block-button"><a class="wp-block-button__link wp-element-button">Button</a></div>
<!-- /wp:button -->

<!-- wp:button -->
<div class="wp-block-button"><a class="wp-block-button__link wp-element-button">Button</a></div>
<!-- /wp:button -->

<!-- wp:button -->
<div class="wp-block-button"><a class="wp-block-button__link wp-element-button">Button</a></div>
<!-- /wp:button -->

<!-- wp:button -->
<div class="wp-block-button"><a class="wp-block-button__link wp-element-button">Button</a></div>
<!-- /wp:button -->

<!-- wp:button -->
<div class="wp-block-button"><a class="wp-block-button__link wp-element-button">Button</a></div>
<!-- /wp:button -->

<!-- wp:button -->
<div class="wp-block-button"><a class="wp-block-button__link wp-element-button">Button</a></div>
<!-- /wp:button -->

<!-- wp:button -->
<div class="wp-block-button"><a class="wp-block-button__link wp-element-button">Button</a></div>
<!-- /wp:button -->

<!-- wp:button -->
<div class="wp-block-button"><a class="wp-block-button__link wp-element-button">Button</a></div>
<!-- /wp:button -->

<!-- wp:button -->
<div class="wp-block-button"><a class="wp-block-button__link wp-element-button">Button</a></div>
<!-- /wp:button -->

<!-- wp:button -->
<div class="wp-block-button"><a class="wp-block-button__link wp-element-button">Button</a></div>
<!-- /wp:button -->

<!-- wp:button -->
<div class="wp-block-button"><a class="wp-block-button__link wp-element-button">Button</a></div>
<!-- /wp:button --></div>
<!-- /wp:buttons -->

<!-- wp:buttons {"layout":{"type":"flex","justifyContent":"left"},"style":{"spacing":{"blockGap":"5px"}}} -->
<div class="wp-block-buttons"><!-- wp:button -->
<div class="wp-block-button"><a class="wp-block-button__link wp-element-button">Button A</a></div>
<!-- /wp:button -->

<!-- wp:button -->
<div class="wp-block-button"><a class="wp-block-button__link wp-element-button">Button B</a></div>
<!-- /wp:button -->

<!-- wp:button -->
<div class="wp-block-button"><a class="wp-block-button__link wp-element-button">Button C</a></div>
<!-- /wp:button --></div>
<!-- /wp:buttons -->

<!-- wp:buttons {"layout":{"type":"flex","justifyContent":"center"},"style":{"spacing":{"blockGap":"2px"}}} -->
<div class="wp-block-buttons"><!-- wp:button -->
<div class="wp-block-button"><a class="wp-block-button__link wp-element-button">Button A</a></div>
<!-- /wp:button -->

<!-- wp:button -->
<div class="wp-block-button"><a class="wp-block-button__link wp-element-button">Button B</a></div>
<!-- /wp:button -->

<!-- wp:button -->
<div class="wp-block-button"><a class="wp-block-button__link wp-element-button">Button C</a></div>
<!-- /wp:button --></div>
<!-- /wp:buttons -->

<!-- wp:buttons {"layout":{"type":"flex","justifyContent":"right"},"style":{"spacing":{"blockGap":"2px"}}} -->
<div class="wp-block-buttons"><!-- wp:button -->
<div class="wp-block-button"><a class="wp-block-button__link wp-element-button">Button A</a></div>
<!-- /wp:button -->

<!-- wp:button -->
<div class="wp-block-button"><a class="wp-block-button__link wp-element-button">Button B</a></div>
<!-- /wp:button -->

<!-- wp:button -->
<div class="wp-block-button"><a class="wp-block-button__link wp-element-button">Button C</a></div>
<!-- /wp:button --></div>
<!-- /wp:buttons -->

<!-- wp:buttons {"layout":{"type":"flex","justifyContent":"center"},"style":{"spacing":{"blockGap":"2px"}}} -->
<div class="wp-block-buttons"><!-- wp:button -->
<div class="wp-block-button"><a class="wp-block-button__link wp-element-button">Button A</a></div>
<!-- /wp:button -->

<!-- wp:button -->
<div class="wp-block-button"><a class="wp-block-button__link wp-element-button">Button B<br>on multiple rows</a></div>
<!-- /wp:button -->

<!-- wp:button -->
<div class="wp-block-button"><a class="wp-block-button__link wp-element-button">Button C</a></div>
<!-- /wp:button --></div>
<!-- /wp:buttons -->

<!-- wp:buttons {"layout":{"type":"flex","justifyContent":"center","verticalAlignment":"top"},"style":{"spacing":{"blockGap":"2px"}}} -->
<div class="wp-block-buttons"><!-- wp:button -->
<div class="wp-block-button"><a class="wp-block-button__link wp-element-button">Button A</a></div>
<!-- /wp:button -->

<!-- wp:button -->
<div class="wp-block-button"><a class="wp-block-button__link wp-element-button">Button B<br>on multiple rows</a></div>
<!-- /wp:button -->

<!-- wp:button -->
<div class="wp-block-button"><a class="wp-block-button__link wp-element-button">Button C</a></div>
<!-- /wp:button --></div>
<!-- /wp:buttons -->

<!-- wp:buttons {"layout":{"type":"flex","justifyContent":"center","verticalAlignment":"bottom"},"style":{"spacing":{"blockGap":"2px"}}} -->
<div class="wp-block-buttons"><!-- wp:button -->
<div class="wp-block-button"><a class="wp-block-button__link wp-element-button">Button A</a></div>
<!-- /wp:button -->

<!-- wp:button -->
<div class="wp-block-button"><a class="wp-block-button__link wp-element-button">Button B<br>on multiple rows</a></div>
<!-- /wp:button -->

<!-- wp:button -->
<div class="wp-block-button"><a class="wp-block-button__link wp-element-button">Button C</a></div>
<!-- /wp:button --></div>
<!-- /wp:buttons -->

<!-- wp:buttons {"layout":{"type":"flex","justifyContent":"center","orientation":"vertical","flexWrap":"wrap"},"style":{"spacing":{"blockGap":"2px"}}} -->
<div class="wp-block-buttons"><!-- wp:button -->
<div class="wp-block-button"><a class="wp-block-button__link wp-element-button">Button A</a></div>
<!-- /wp:button -->

<!-- wp:button -->
<div class="wp-block-button"><a class="wp-block-button__link wp-element-button">Button B</a></div>
<!-- /wp:button -->

<!-- wp:button -->
<div class="wp-block-button"><a class="wp-block-button__link wp-element-button">Button C</a></div>
<!-- /wp:button --></div>
<!-- /wp:buttons -->
```

</details>

<img width="814" alt="Screen Shot 2022-07-25 at 11 27 27 am" src="https://user-images.githubusercontent.com/6458278/180676720-8fd9ac9a-a2f7-4cd5-b189-5a4f3539e7f0.png">

Enable `"useRootPaddingAwareAlignments": true` in your theme.json and give a group block full alignment and some left and right padding. 

The layout styles should be printed as expected, e.g., `.wp-block-group.wp-container-6 > .alignfull {margin-right: calc(26px * -1); margin-left: calc(69px * -1);}`

Also make sure the CSS rules are being batched into a single, inline style sheet. For example, this:

```html
<style id='block-supports-inline-css'>
.wp-elements-bb563b31fce88a8bb12898100773d8e7 a {color: #842ecf;}.wp-elements-4cebe23f336adf33d0fd72ee6511e88e a {color: #fbc941;}.wp-elements-58f56fce71e24389540d1fff37a68db8 a,.wp-elements-cad477801df51349bcb7ad151a4593cb a {color: var(--wp--preset--color--pale-pink);}
</style>
<style id='layout-block-supports-inline-css'>
.wp-block-group.wp-container-6 > .alignfull {margin-right: calc(26px * -1); margin-left: calc(69px * -1);}.wp-block-group.wp-container-10 > :where(:not(.alignleft):not(.alignright):not(.alignfull)) {max-width: 402px; margin-left: auto !important; margin-right: auto !important;}.wp-block-group.wp-container-10 > .alignwide {max-width: 402px;}.wp-block-column.wp-container-3 > :where(:not(.alignleft):not(.alignright):not(.alignfull)),.wp-block-column.wp-container-7 > :where(:not(.alignleft):not(.alignright):not(.alignfull)),.wp-block-column.wp-container-11 > :where(:not(.alignleft):not(.alignright):not(.alignfull)) {max-width: 26px; margin-left: auto !important; margin-right: auto !important;}.wp-block-column.wp-container-3 > .alignwide,.wp-block-column.wp-container-7 > .alignwide,.wp-block-column.wp-container-11 > .alignwide {max-width: 26px;}.wp-block-column.wp-container-3 .alignfull,.wp-block-group.wp-container-6 .alignfull,.wp-block-column.wp-container-7 .alignfull,.wp-block-group.wp-container-10 .alignfull,.wp-block-column.wp-container-11 .alignfull,.wp-block-post-content.wp-container-16 .alignfull {max-width: none;}.wp-block-columns.wp-container-5,.wp-block-columns.wp-container-9,.wp-block-columns.wp-container-13,.wp-block-group.wp-container-14 {flex-wrap: nowrap;}.wp-block-group.wp-container-6 > :where(:not(.alignleft):not(.alignright):not(.alignfull)),.wp-block-post-content.wp-container-16 > :where(:not(.alignleft):not(.alignright):not(.alignfull)) {max-width: 840px; margin-left: auto !important; margin-right: auto !important;}.wp-block-group.wp-container-6 > .alignwide,.wp-block-post-content.wp-container-16 > .alignwide {max-width: 1100px;}
</style>
```

Instead of this:

```html
	<style>.wp-block-navigation.wp-container-3 {justify-content: flex-end;}</style>
<style>.wp-block-group.wp-container-4 {justify-content: space-between;}</style>
<style>.wp-block-group.wp-container-5 > :where(:not(.alignleft):not(.alignright):not(.alignfull)) {max-width: 650px;margin-left: auto !important;margin-right: auto !important;}.wp-block-group.wp-container-5 > .alignwide { max-width: 1000px;}.wp-block-group.wp-container-5 .alignfull { max-width: none; }</style>
<style>.wp-block-group.wp-container-6 > :where(:not(.alignleft):not(.alignright):not(.alignfull)) {max-width: 650px;margin-left: auto !important;margin-right: auto !important;}.wp-block-group.wp-container-6 > .alignwide { max-width: 1000px;}.wp-block-group.wp-container-6 .alignfull { max-width: none; }</style>
<style>.wp-elements-bb563b31fce88a8bb12898100773d8e7 a { color: #842ecf; }</style>
<style>.wp-elements-bb563b31fce88a8bb12898100773d8e7 a { color: #842ecf; }</style>
<style>.wp-elements-58f56fce71e24389540d1fff37a68db8 a { color: var(--wp--preset--color--pale-pink); }</style>
<style>.wp-elements-58f56fce71e24389540d1fff37a68db8 a { color: var(--wp--preset--color--pale-pink); }</style>
<style>.wp-elements-cad477801df51349bcb7ad151a4593cb a { color: var(--wp--preset--color--pale-pink); }</style>
<style>.wp-elements-4cebe23f336adf33d0fd72ee6511e88e a { color: #fbc941; }</style>
<style>.wp-elements-4cebe23f336adf33d0fd72ee6511e88e a { color: #fbc941; }</style>
<style>.wp-block-column.wp-container-9 > :where(:not(.alignleft):not(.alignright):not(.alignfull)) {max-width: 26px;margin-left: auto !important;margin-right: auto !important;}.wp-block-column.wp-container-9 > .alignwide { max-width: 26px;}.wp-block-column.wp-container-9 .alignfull { max-width: none; }</style>
<style>.wp-block-columns.wp-container-11 { flex-wrap: nowrap; }</style>
<style>.wp-block-group.wp-container-12 > :where(:not(.alignleft):not(.alignright):not(.alignfull)) {max-width: 650px;margin-left: auto !important;margin-right: auto !important;}.wp-block-group.wp-container-12 > .alignwide { max-width: 1000px;}.wp-block-group.wp-container-12 .alignfull { max-width: none; }</style>
<style>.wp-block-column.wp-container-13 > :where(:not(.alignleft):not(.alignright):not(.alignfull)) {max-width: 26px;margin-left: auto !important;margin-right: auto !important;}.wp-block-column.wp-container-13 > .alignwide { max-width: 26px;}.wp-block-column.wp-container-13 .alignfull { max-width: none; }</style>
<style>.wp-block-columns.wp-container-15 { flex-wrap: nowrap; }</style>
<style>.wp-block-group.wp-container-16 > :where(:not(.alignleft):not(.alignright):not(.alignfull)) {max-width: 402px;margin-left: auto !important;margin-right: auto !important;}.wp-block-group.wp-container-16 > .alignwide { max-width: 402px;}.wp-block-group.wp-container-16 .alignfull { max-width: none; }</style>
<style>.wp-block-column.wp-container-17 > :where(:not(.alignleft):not(.alignright):not(.alignfull)) {max-width: 26px;margin-left: auto !important;margin-right: auto !important;}.wp-block-column.wp-container-17 > .alignwide { max-width: 26px;}.wp-block-column.wp-container-17 .alignfull { max-width: none; }</style>
<style>.wp-block-columns.wp-container-19 { flex-wrap: nowrap; }</style>
<style>.wp-block-group.wp-container-20 { flex-wrap: nowrap; }</style>
<style>.wp-block-post-content.wp-container-22 > :where(:not(.alignleft):not(.alignright):not(.alignfull)) {max-width: 650px;margin-left: auto !important;margin-right: auto !important;}.wp-block-post-content.wp-container-22 > .alignwide { max-width: 1000px;}.wp-block-post-content.wp-container-22 .alignfull { max-width: none; }</style>
<style>.wp-block-group.wp-container-24 > :where(:not(.alignleft):not(.alignright):not(.alignfull)) {max-width: 650px;margin-left: auto !important;margin-right: auto !important;}.wp-block-group.wp-container-24 > .alignwide { max-width: 1000px;}.wp-block-group.wp-container-24 .alignfull { max-width: none; }</style>
<style>.wp-block-group.wp-container-26 {justify-content: space-between;}</style>
<style>.wp-block-group.wp-container-27 > :where(:not(.alignleft):not(.alignright):not(.alignfull)) {max-width: 650px;margin-left: auto !important;margin-right: auto !important;}.wp-block-group.wp-container-27 > .alignwide { max-width: 1000px;}.wp-block-group.wp-container-27 .alignfull { max-width: none; }</style>
<style>.wp-block-group.wp-container-28 > :where(:not(.alignleft):not(.alignright):not(.alignfull)) {max-width: 650px;margin-left: auto !important;margin-right: auto !important;}.wp-block-group.wp-container-28 > .alignwide { max-width: 1000px;}.wp-block-group.wp-container-28 .alignfull { max-width: none; }</style>

```

ramonjd committed:
Enqueuing block support styles version 1.

ramonjd committed:
Linter, this one's for you.

ramonjd committed:
Post trunk merge cleanup and update tests.

ramonjd commented on packages/style-engine/class-wp-style-engine.php:
```diff
@@ -236,6 +255,67 @@ public static function get_instance() {
+				continue;
+			}
+
+			$css_rules = $store_instance->get_all_rules();
```
This will eventually be updated when https://github.com/WordPress/gutenberg/pull/42463 gets in

andrewserong replied:
Ah, you already answered my question from my review (https://github.com/WordPress/gutenberg/pull/42452#pullrequestreview-1041223364). Cheers! 🍺

ramonjd committed:
Removed spacing around curly braces in CSS rules. Updated tests.
We could maybe add a prettify option down the road.

Juggling methods around to cater for adding styles to (any) store.
Also making return values consistent.

andrewserong commented on packages/style-engine/class-wp-style-engine.php:
```diff
@@ -236,6 +255,67 @@ public static function get_instance() {
+	 * Fetches, processes and compiles stored styles, then renders them to the page.
+	 */
+	public static function enqueue_rendered_styles() {
+		foreach ( static::$stores as $store_key => $store_instance ) {
```
Will we need to check each store as to whether or not it should be enqueued (as a flag on the store, or something like that)? Or to put it a different way, what if we want to enqueue block supports styles, but _not_ enqueue global styles? My thinking here is that to iteratively start using the style engine in global styles, we might want to (potentially) add styles to the `global-styles` store, but output those styles as a string to be concatenated to the global styles rules, instead of being enqueued separately.

Of course, there's nothing preventing us from doing that really incrementally in global styles without the `'enqueue'` option, so feel free to disregard if you think this is a non-issue!

ramonjd replied:
> Will we need to check each store as to whether or not it should be enqueued (as a flag on the store, or something like that)? Or to put it a different way, what if we want to enqueue block supports styles, but not enqueue global styles? 

Yes we will! 😄 Thanks for the nudge there. I had something in mind, but I've just got to remember it.

ramonjd replied:
I've solved this by splitting the compile and parse functionality and creating two public methods 👍

andrewserong commented on packages/style-engine/class-wp-style-engine.php:
```diff
@@ -236,6 +255,67 @@ public static function get_instance() {
+
+			$styles_output = '';
+
+			foreach ( $css_rules as $selector => $css_rule ) { // phpcs:ignore VariableAnalysis.CodeAnalysis.VariableAnalysis.UnusedVariable
```
Tiny nit: can we avoid the `phpcs:ignore` line by removing `$selector` from the `foreach`, or is it still needed for follow-ups?

```suggestion
			foreach ( $css_rules as $css_rule ) {
```

ramonjd replied:
Defo. Thanks. That was me trying to check my work in on a Friday afternoon 😄

andrewserong commented on packages/style-engine/class-wp-style-engine.php:
```diff
@@ -523,3 +610,32 @@ function wp_style_engine_get_block_supports_styles( $block_styles, $options = ar
+ *     'classnames'    => (string) Classnames separated by a space.
+ * );
+ */
+function wp_style_engine_enqueue_block_supports_styles( $selector, $block_styles ) {
```
I like the additional public method for this, as it makes it a bit clearer than adding additional options to the `wp_style_engine_get_block_supports_styles` function. That said, given we have an `$options` param on the other function, should we try to somehow consolidate the two, or do you think it's better having the two separate public functions for better clarity?

I suppose my idle thought is: is it better to have a single public function with lots of options, or separate functions that are simpler to work with, but we then have more public functions in the global namespace 🤔

ramonjd replied:
> is it better to have a single public function with lots of options, or separate functions that are simpler to work with, but we then have more public functions in the global namespace

This is the struggle. 😄 

The reason I went with two was to separate the concerns (at least ostensibly) and provide a declarative way of enqueuing block styles so the consumer wouldn't have to worry about the underlying implementation. 

That `wp_style_engine_get_block_supports_styles` accepts the enqueue option does mix concerns in that way, and I think it'd be cleaner to somehow separate parsing from enqueuing.

Another thing I should keep in mind are bespoke styles we'll want to enqueue such as those for layout and global styles.

I'm not sure. I'll try some things out.

ramonjd replied:
So I might have gone over the top a bit, but I was trying to separate out parsing methods from compilation (to CSS strings etc)

Just testing things out.

Wondering there's a case for more abstraction, e.g., pulling out parsing methods from the main class.

andrewserong reviewed (commented):
Nice one @ramonjd, I like the direction this is taking, just added a few comments. And it looks like overall it'll be quite compatible with the Renderer  / Optimizer approach in #42463 — once that one lands, we'd switch over to using that in `enqueue_rendered_styles`?

ramonjd commented on packages/style-engine/class-wp-style-engine.php:
```diff
@@ -293,6 +287,159 @@ protected static function is_valid_style_value( $style_value ) {
+	 *
+	 * @return string A compiled CSS string.
+	 */
+	public function store_css_rule( $css_selector, $css_declarations, $store_key ) {
```
This could be the public method that layouts might use, or any other area that wants to store non-block-support-type styles.

ramonjd committed:
Splitting `wp_style_engine_enqueue_block_supports_styles` and `wp_style_engine_get_block_supports_styles` so we can enqueue styles that don't need parsing, e.g., layout

ramonjd commented on packages/style-engine/class-wp-style-engine.php:
```diff
@@ -507,19 +625,67 @@ protected static function get_individual_property_css_declarations( $style_value
+ *
+ * @return boolean Whether the storage process was successful.
+ */
+function wp_style_engine_enqueue_block_supports_styles( $selector, $css_declarations ) {
```
So this is the method I was thinking layout could use, given that the layout styles don't need any block styles parsing.

ramonjd commented on packages/style-engine/class-wp-style-engine.php:
```diff
@@ -293,6 +287,159 @@ protected static function is_valid_style_value( $style_value ) {
+	 *     'classnames'       => (array) A flat array of classnames.
+	 * );
+	 */
+	public function parse_block_supports_styles( $block_styles, $options ) {
```
Now that the compile and parse functionality is split, it might be a change to move all the parsing-related methods and utilities to another class or file.

ramonjd committed:
Integrate the processor class

ramonjd committed:
Migrate layout styles to style engine store.

ramonjd commented on packages/style-engine/class-wp-style-engine.php:
```diff
@@ -293,6 +288,158 @@ protected static function is_valid_style_value( $style_value ) {
+	/**
+	 * Fetches, processes and compiles stored styles, then renders them to the page.
+	 */
+	public static function process_and_enqueue_stored_styles() {
```
The enqueue and WP hook logic could well live in Gutenberg itself, e.g., `script-loader.php` with the style engine only taking care of `parse + store + compile`

aristath commented on packages/style-engine/class-wp-style-engine.php:
```diff
@@ -507,19 +642,92 @@ protected static function get_individual_property_css_declarations( $style_value
  * );
  */
 function wp_style_engine_get_block_supports_styles( $block_styles, $options = array() ) {
+	if ( class_exists( 'WP_Style_Engine' ) ) {
+		$defaults = array(
+			'selector'                   => null,
+			'convert_vars_to_classnames' => false,
+			'enqueue'                    => false,
+		);
+
+		$options       = wp_parse_args( $options, $defaults );
+		$style_engine  = WP_Style_Engine::get_instance();
+		$parsed_styles = $style_engine->parse_block_supports_styles( $block_styles, $options );
+
+		// Output.
+		$styles_output = array();
+		if ( ! empty( $parsed_styles['css_declarations'] ) ) {
+			$styles_output['css']          = $style_engine->compile_css( $parsed_styles['css_declarations'], $options['selector'] );
+			$styles_output['declarations'] = $parsed_styles['css_declarations'];
+			if ( true === $options['enqueue'] ) {
+				$style_engine::store_css_rule( $options['selector'], $parsed_styles['css_declarations'], 'block-supports' );
+			}
+		}
+
+		if ( ! empty( $parsed_styles['classnames'] ) ) {
+			$styles_output['classnames'] = $style_engine->compile_classnames( $parsed_styles['classnames'] );
+		}
+
+		return array_filter( $styles_output );
+	}
+	return array();
```
Just a nitpick, to reduce the depth of indentations: Instead of wrapping the whole code in `if class_exists()`, we can return early if the class does not exist 👍 
```suggestion
	if ( ! class_exists( 'WP_Style_Engine' ) ) {
		return array();
	}
	$defaults = array(
		'selector'                   => null,
		'convert_vars_to_classnames' => false,
		'enqueue'                    => false,
	);

	$options       = wp_parse_args( $options, $defaults );
	$style_engine  = WP_Style_Engine::get_instance();
	$parsed_styles = $style_engine->parse_block_supports_styles( $block_styles, $options );

	// Output.
	$styles_output = array();
	if ( ! empty( $parsed_styles['css_declarations'] ) ) {
		$styles_output['css']          = $style_engine->compile_css( $parsed_styles['css_declarations'], $options['selector'] );
		$styles_output['declarations'] = $parsed_styles['css_declarations'];
		if ( true === $options['enqueue'] ) {
			$style_engine::store_css_rule( $options['selector'], $parsed_styles['css_declarations'], 'block-supports' );
		}
	}

	if ( ! empty( $parsed_styles['classnames'] ) ) {
		$styles_output['classnames'] = $style_engine->compile_classnames( $parsed_styles['classnames'] );
	}

	return array_filter( $styles_output );
```

ramonjd replied:
Excellent. Committed!

ramonjd committed:
Update packages/style-engine/class-wp-style-engine.php

Co-authored-by: Ari Stathopoulos <aristath@gmail.com>

aristath committed:
Tweaks for #42452 (#42691)

* abstract stores

* "else" not needed

* compile_classnames method not needed

* we have a method to get the store here

* Make the wp_style_engine_add_to_store function always return a store

* wp_style_engine_get_stylesheet - always return string

* Merged with base branch.
Added test for new method.

Co-authored-by: ramonjd <ramonjd@gmail.com>
Co-authored-by: Ramon <ramonjd@users.noreply.github.com>

aristath commented on packages/style-engine/class-wp-style-engine.php:
```diff
@@ -507,19 +607,91 @@ protected static function get_individual_property_css_declarations( $style_value
+		return '';
 	}
-	return null;
+	return WP_Style_Engine::get_instance()::compile_stylesheet_from_store( $store_key );
```
I believe we don't need to get the instance here, right?

```suggestion
	return WP_Style_Engine::compile_stylesheet_from_store( $store_key );
```

aristath reviewed (approved):
I've been looking at this PR and testing it for a couple of days... Overall, it looks good. Should we merge it? If there's anything else, it can be done in follow-up PRs. The code here looks good, and it works fine.
It would be good to get this one in so we can continue implementing the style-engine in other places too 👍

ramonjd committed:
Adding check for the context argument.
Adding tests.

andrewserong commented on lib/block-supports/layout.php:
```diff
@@ -136,29 +141,35 @@ function gutenberg_get_layout_style( $selector, $layout, $has_block_gap_support
+
+		// Return compiled layout styles to retain backwards compatibility.
+		// Since https://github.com/WordPress/gutenberg/pull/42452 we no longer call wp_enqueue_block_support_styles in this block supports file.
+		return gutenberg_style_engine_get_stylesheet( 'layout-block-supports' );
```
Oh, this is a tricky one — the return here is cumulative in that the output appears to contain all previously registered layout styles for that store.

If I log out this return value, here's how it looks in console output (note that the first line is a single rule, and the subsequent lines are that rule plus the next instance of the rendering):

<img width="1196" alt="image" src="https://user-images.githubusercontent.com/14988353/181427813-4f10a942-2e01-4dc7-89b9-315394017c5f.png">

ramonjd replied:
> Oh, this is a tricky one — the return here is cumulative in that the output appears to contain all previously registered layout styles for that store.

Yeah, I think we might have to bite the bullet and manually output them _in situ_

aristath replied:
In my other PR I had added a `$cleanup_rules` in the `get_css()` method to avoid this issue. See https://github.com/WordPress/gutenberg/pull/42592/files#diff-f23bfb067de6325fe30b9ec68f807342a3286e40b87937bcb806ea285e0252a3R44
Basically whenever CSS was getting printed by the `WP_Style_Engine_Processor` class, I had the option to pass `true` as an arg to the method and "reset" the stores so there's no duplication of styles.

ramonjd replied:
> Basically whenever CSS was getting printed by the WP_Style_Engine_Processor class, I had the option to pass true as an arg to the method and "reset" the stores so there's no duplication of styles.

Interesting, thanks! 

Sounds like it could help us in this situation: we only want to return the styles that are collected in the body of the layout function during a single execution. What we do for backward compatibility! 😄 

I haven't tested so could be way wrong here, but I think the snag is that nothing is rendered yet.  We're relying on the full contents of the store existing later when enqueuing.

andrewserong commented on lib/block-supports/layout.php:
```diff
@@ -74,12 +73,12 @@ function gutenberg_get_layout_style( $selector, $layout, $has_block_gap_support
-					$padding_left = $block_spacing_values['declarations']['padding-left'];
-					$style       .= "$selector > .alignfull { margin-left: calc($padding_left * -1); }";
+					$padding_left                              = $block_spacing_values['declarations']['padding-left'];
+					$layout_styles[ "$selector > .alignfull" ] = array( 'margin-left' => "calc($padding_left * -1)" );
```
This line overrides the `margin-right` rule so that only `margin-left` is applied.

| Trunk | This PR |
| --- | --- |
| <img width="429" alt="image" src="https://user-images.githubusercontent.com/14988353/181429495-eed0be18-8056-4021-a8a4-b89c8d7050f7.png"> | <img width="428" alt="image" src="https://user-images.githubusercontent.com/14988353/181429573-996312da-0bca-4f7b-a4aa-668bda8c455d.png"> |

ramonjd replied:
🤦 Thanks for catching this stuff. 

I'll look into a different way to assemble the styles here.

aristath replied:
@ramonjd I just pushed a commit fixing these issues 👍

aristath replied:
reverted fix... I rushed into it and did it wrong

andrewserong commented on lib/block-supports/layout.php:
```diff
@@ -123,9 +130,7 @@ function gutenberg_get_layout_style( $selector, $layout, $has_block_gap_support
-				$style .= "$selector {";
-				$style .= "gap: $gap_value;";
-				$style .= '}';
+				$layout_styles[ $selector ] = array( 'gap' => $gap_value );
```
This rule is overwriting the `flex-wrap` rule set above.

aristath replied:
fixed

aristath replied:
reverted fix... I rushed into it and did it wrong

andrewserong commented on lib/block-supports/layout.php:
```diff
@@ -136,29 +141,35 @@ function gutenberg_get_layout_style( $selector, $layout, $has_block_gap_support
-				$style .= "$selector {";
-				$style .= "justify-content: {$justify_content_options[ $layout['justifyContent'] ]};";
-				$style .= '}';
+				$layout_styles[ $selector ] = array( 'justify-content' => $justify_content_options[ $layout['justifyContent'] ] );
```
As above, this overwrites the gap rule.

aristath replied:
fixed

aristath replied:
reverted fix... I rushed into it and did it wrong

andrewserong commented on lib/block-supports/layout.php:
```diff
@@ -136,29 +141,35 @@ function gutenberg_get_layout_style( $selector, $layout, $has_block_gap_support
-				$style .= "$selector {";
-				$style .= "align-items: {$vertical_alignment_options[ $layout['verticalAlignment'] ]};";
-				$style .= '}';
+				$layout_styles[ $selector ] = array( 'align-items' => $vertical_alignment_options[ $layout['verticalAlignment'] ] );
```
As above, this overwrites the gap rule.

aristath replied:
fixed

aristath replied:
reverted fix... I rushed into it and did it wrong

andrewserong commented on lib/block-supports/layout.php:
```diff
@@ -136,29 +141,35 @@ function gutenberg_get_layout_style( $selector, $layout, $has_block_gap_support
-			$style .= 'flex-direction: column;';
+			$layout_styles[ $selector ] = array(
+				'flex-direction' => 'column',
+				'align-items'    => 'flex-start',
+			);
 			if ( ! empty( $layout['justifyContent'] ) && array_key_exists( $layout['justifyContent'], $justify_content_options ) ) {
-				$style .= "align-items: {$justify_content_options[ $layout['justifyContent'] ]};";
+				$layout_styles[ $selector ] = array( 'align-items' => $justify_content_options[ $layout['justifyContent'] ] );
 			} else {
-				$style .= 'align-items: flex-start;';
+				$layout_styles[ $selector ] = array( 'align-items' => 'flex-start' );
```
As above, this overwrites the gap rule.

aristath replied:
fixed

aristath replied:
reverted fix... I rushed into it and did it wrong

andrewserong reviewed (commented):
Nice work @ramonjd and @aristath, this looks like it's getting much closer! Nice addition with including a default block-supports context when enqueuing 👍

I ran into a couple of issues while testing the Layout support. The main issue is that it looks like properties with the same selector are being overwritten instead of being concatenated. I've added comments on each of the features I noticed that occurring in, but 🤞 that one should be a fairly straightforward fix.

I wasn't too sure how to go about solving the fallback output for Layout styles, though? I've left a comment on that one.

Happy to do more testing tomorrow!

Update: for ease of testing, I've added some test Layout markup to the PR description. Hopefully it should be pretty clear what the expected result should be when switching between the post editor view and the site front-end.

ramonjd committed:
Updating the processor so that it's ignorant of stores. Why? So that it can be used to process any CSS and not just stored CSS.
Updating layout for backwards compatibility in gutenberg_get_layout_style (returning the styles that are collected in the function body only)
Created a new mode for incoming $css_rules to be a collection of selector + css_declaration keys.
Removed wp_style_engine_get_stylesheet (from store) since we don't use it yet
Added a new function wp_style_engine_get_stylesheet_from_css_rules() that will process and compile a collection of CSS rules, and not store them.

ramonjd committed:
dump var_dump()

aristath committed:
Improve the processor

aristath committed:
remove trailing commas - compatibility with PHP < 7.2

aristath committed:
rename css_declarations to declarations

aristath committed:
remove unused variable

Removing unused function wp_style_engine_get_stylesheet_from_store
Updating tests to check for merging and deduping

andrewserong commented on lib/block-supports/layout.php:
```diff
@@ -136,29 +171,47 @@ function gutenberg_get_layout_style( $selector, $layout, $has_block_gap_support
+
+		// Return compiled layout styles to retain backwards compatibility.
+		// Since https://github.com/WordPress/gutenberg/pull/42452 we no longer call wp_enqueue_block_support_styles in this block supports file.
+		return gutenberg_style_engine_get_stylesheet_from_css_rules( $layout_styles );
```
Nice one, this looks like a good way to abstract it to me (a clean public function, and that function calls a simple method on the Style Engine class, which in turn calls the Processor class 👍).

Output looks good via logging, each rendered block gets its own string:

<img width="1210" alt="image" src="https://user-images.githubusercontent.com/14988353/182062767-06659bf9-a743-4f18-a6f9-22fc4f8bafe7.png">

ramonjd committed:
Switch parse_block_styles from public to protected static

ramonjd committed:
Now that all methods are static, there's no need to call `get_instance()`

ramonjd committed:
Revert get_instance() in wp_style_engine_add_to_store because we want to ensure the hook is enqueued here.

andrewserong commented on packages/style-engine/class-wp-style-engine-processor.php:
```diff
@@ -12,26 +12,52 @@
+	 * @param WP_Style_Engine_CSS_Rules_Store $store The store to add.
+	 */
+	public function add_store( WP_Style_Engine_CSS_Rules_Store $store ) {
+		$this->stores[] = $store;
```
What happens if we call `add_store` twice with the same store? Should the stores be named, as they are in `WP_Style_Engine_CSS_Rules_Store` so we avoid duplication? Or should the Processor not really care about knowing anything about the stores 🤔

ramonjd replied:
> Or should the Processor not really care about knowing anything about the stores 

I would guess "yes" to this, and I suppose it's less of a risk since we instantiate `WP_Style_Engine_Processor` when we want to use it. 

I suppose it wouldn't be much of a change to allow for keys though. 

Interested to hear @aristath's take on it.

andrewserong replied:
I'm happy either way with this one, just thought I'd ask 🙂

aristath replied:
You're right, it would make sense to have them named here. I pushed a change and now there won't be duplicates 👍

andrewserong commented on packages/style-engine/class-wp-style-engine.php:
```diff
@@ -293,6 +278,154 @@ protected static function is_valid_style_value( $style_value ) {
+	 */
+	private function __construct() {
+		// Register the hook callback to render stored styles to the page.
+		static::render_styles( array( __CLASS__, 'process_and_enqueue_stored_styles' ) );
```
In terms of when we call this, at the moment the registration appears to happen when `get_instance()` happens to first be called in the code.

I was wondering if it's worth exploring using an explicit action for registration, so that it's predictable when the registration happens? I can't currently imagine a particular circumstance where this would be an issue, but something on my mind is the bug we fixed in https://github.com/WordPress/gutenberg/pull/42756 which was a difficult to diagnose one due to not really knowing when something was called.

I don't think this is a blocker for this PR, as it'd be easy enough for us to try out a different place for registration in a follow-up, but just thought I'd jot down the thought!

ramonjd replied:
> I was wondering if it's worth exploring using an explicit action for registration, so that it's predictable when the registration happens? 

It's a great question.

I was actually wondering whether the style engine should be enqueuing scripts/registering hooks at all, or whether Gutenberg (the consumer) should take care of its own registration / callbacks, and decides which stores it prints. 

I agree that it's not very clear how this PR registers the hook. The constraint was I wanted to only register it once, so  it was the easiest place to put it.

andrewserong replied:
> The constraint was I wanted to only register it once, so it was the easiest place to put it.

Exactly, I think that makes for a good argument for keeping it there for now, and we can look at where an ideal place is potentially in a follow-up?

ramonjd replied:
> something on my mind is the bug we fixed in https://github.com/WordPress/gutenberg/pull/42756 which was a difficult to diagnose one due to not really knowing when something was called.

Yeah, I've made it fairly opaque. It's not clear that `WP_Style_Engine::get_instance();` needs to be called for the styles to be enqueued.

That needs to be fixed I reckon, otherwise it might create a huge rabbit hole.

andrewserong commented on packages/style-engine/class-wp-style-engine.php:
```diff
@@ -293,6 +278,154 @@ protected static function is_valid_style_value( $style_value ) {
+	 *
+	 * @see gutenberg_enqueue_block_support_styles()
+	 */
+	protected static function render_styles( $callable, $priority = 10 ) {
```
Just another idle thought, but `render_styles` makes me think of an imperative action (calling the function will immediately do something), whereas the logic in this function is a little more subtle in that it's registering callbacks to be fired later. I was wondering if something like `register_actions` might be closer to what the function's doing? Again, not a blocker for this PR, as it's a private function.

andrewserong commented on packages/style-engine/class-wp-style-engine.php:
```diff
@@ -511,21 +614,127 @@ protected static function get_individual_property_css_declarations( $style_value
+	}
+
+	// Get instance here to ensure that we register hooks to enqueue stored styles.
+	$style_engine = WP_Style_Engine::get_instance();
```
Should the hooks be registered explicitly somewhere in WordPress, rather than when first called? Or, is it safe to leave it when this is called so that it's only registered if it's needed 🤔

This is mostly an idle thought and linked to another comment, so I don't think of it as a blocker to this PR, it's definitely something we can look at in a separate follow-up if need be.

ramonjd replied:
> Should the hooks be registered explicitly somewhere in WordPress, rather than when first called? 

Do you mean outside the style engine? As [mentioned above](https://github.com/WordPress/gutenberg/pull/42452#discussion_r934145324), I get the inkling that it might be better to give the consumer control over how styles are rendered via hooks / enqueues etc.

I think it's worth exploring in a follow up. It might mean moving a bit of logic to `script-loader.php`... not sure. 


> Or, is it safe to leave it when this is called so that it's only registered if it's needed

My thinking was that, as soon as at least one store is created, then we know to register the hook. 

Whether it's safe or not in the long term I'm not so sure, but if you have doubts then it probably won't be 😄

andrewserong replied:
> Do you mean outside the style engine?

Potentially, yes. Once the style engine is merged into core, we'll need Gutenberg to be able to switch off the core version, so whichever way we go about it, we'll need to be clear on the logic. But you raise a great point about giving consumers control over how styles are rendered!

> I think it's worth exploring in a follow up. It might mean moving a bit of logic to script-loader.php... not sure

Absolutely, I think it'll be easier to figure out the desired approach in an isolated PR. Or at the very least, figure out in a separate PR whether or not it's worth going down alternative rabbitholes 😄

> My thinking was that, as soon as at least one store is created, then we know to register the hook.

Personally, I think that's a solid strategy for the style engine — it only registers what it needs. It might be that it really is the best way to go about it, for the style engine to do the registering when it needs to. Either way it'd be good for us to document this as a principle to aid future debugging.

At this stage, I'm not too fussed about which particular direction we go in — the main goal with my comments is to ensure that we've made a clear decision about what we're attempting to do, so we can then use that to guide follow-ups. I quite like the "as soon as at least one store is created, then we know to register the hook." principle, and seeing how far that gets us 👍

aristath replied:
I agree, it would feel better if enqueuing etc happen outside the class. But I think we can do that in a followup PR, it doesn't have to be done all at once 👍

andrewserong commented on lib/block-supports/layout.php:
```diff
@@ -123,9 +157,10 @@ function gutenberg_get_layout_style( $selector, $layout, $has_block_gap_support
-				$style .= '}';
+				$layout_styles[] = array(
+					'selector'     => $selector,
+					'declarations' => array( 'gapp' => $gap_value ),
```
Typo 😄

```suggestion
					'declarations' => array( 'gap' => $gap_value ),
```

ramonjd replied:
`gapp: 10ppx` :trollface: 

Thanks for catching 😄

andrewserong reviewed (commented):
This is coming along nicely! Most of the issues I reported earlier have been resolved, except for a typo in flex `gap` 😄

## Layout support testing:

✅ Returned strings are no longer cumulative 👍

### Default layout

✅ Custom `contentSize`
✅ Custom `wideSize`
✅ Full width alignment
✅ Padding left/right offset in full-width children
✅ Gap values (top-margin)

### Flex layout

✅ Orientation
✅ Horizontal content justification
✅ Vertical content justification
✅ No-wrap setting
❌ Gap styles (looks it's just a typo)

---

I added a few questions / thoughts about registering hooks, but I don't think they're really blockers — they can probably be looked into / explored separately.

Overall, I think this is looking very close!

ramonjd committed:
Adding a test for the 'enqueue' flag.

ramonjd committed:
Adding a test for the 'enqueue' flag.

ramonjd committed:
Update lib/block-supports/layout.php

Co-authored-by: Andrew Serong <14988353+andrewserong@users.noreply.github.com>

ramonjd committed:
Merge branch 'try/style-engine-enqueue-block-supports-styles' of github.com:WordPress/gutenberg into try/style-engine-enqueue-block-supports-styles

ramonjd commented:
Thanks for the detailing testing and feedback @andrewserong! 🙇 

 I'll have a play around with some things in relation to the enqueuing.

Also, I've been playing about with extracting all the parsing functionality to make the main class a little more focussed (as a hub)

https://github.com/WordPress/gutenberg/pull/42843

It's just there for later as a follow up in case folks think it's worth it.

ramonjd commented on packages/style-engine/class-wp-style-engine.php:
```diff
@@ -293,6 +278,154 @@ protected static function is_valid_style_value( $style_value ) {
+	public static function process_and_enqueue_stored_styles() {
+		$stores = WP_Style_Engine_CSS_Rules_Store::get_stores();
+
+		foreach ( $stores as $key => $store ) {
```
If we move this stuff to Gutenberg, then we could be explicit about the stores here and combine them into one `<style />`, e.g., 

```php
$block_supports_store = static::get_store( 'block-supports' );
$block_supports_layout_store = static::get_store( 'layout-block-supports' );
$processor = new WP_Style_Engine_Processor();
$processor->add_store( $block_supports_store );
$processor->add_store( $block_supports_layout_store );
$styles = $processor->get_css();
```

aristath committed:
Add named stores to the processor

aristath committed:
avoid setting var for something that only gets used once

aristath committed:
Only use "else" if absolutely necessary

aristath committed:
Add a set_name method

aristath committed:
combine & simplify conditions

aristath committed:
use empty() instead of isset() checks here

aristath committed:
shorten it

andrewserong reviewed (approved):
Thanks for all the updates and tweaks @ramonjd and @aristath! I've re-tested the Elements support and each of the features of the Layout support, and it appears to be all working nicely now 🎉

The tidied up function names look good to me, and let's continue exploring how we register actions, etc, in follow-ups.

I might be overly cautious here, but prior to merging it'd be good to update the PR description with a bit more context, so that if this does introduce any issues, it's a little clearer what's changed to help folks debugging. It could also be worth exploring us switching off the optimisation step of the Processor in this PR, and then switch it on in a separate PR so that we have different commits in `trunk` for each step — this way, if there are any issues caused by the changes to Layout output, it'll hopefully be easier to identify exactly the step that caused the issue. Totally not a blocker if you think that's overkill though.

I suspect it'll be perfectly fine, but I recently ran into some weird edge cases with Layout, so just thought I'd mention it in case it helps us with the merge process.

This PR is a great milestone, though, it's a really significant improvement to the Layout style output, and IMO delivers on a lot of the promise of introducing a style engine. Kudos! ✨

ramonjd commented:
Thanks @andrewserong and @aristath for all the help getting this one to a mergeable state! 🙇 

> I might be overly cautious here, but prior to merging it'd be good to update the PR description with a bit more context, so that if this does introduce any issues, it's a little clearer what's changed to help folks debugging. 

Will do, thanks!

> It could also be worth exploring us switching off the optimisation step of the Processor in this PR, and then switch it on in a separate PR so that we have different commits in trunk for each step — this way, if there are any issues caused by the changes to Layout output, it'll hopefully be easier to identify exactly the step that caused the issue. Totally not a blocker if you think that's overkill though.

I think this is a pretty good step. I'll merge this one with the combination optimizations turned on, but follow it up quickly with an `'optimize'` flag that we can switch on once we know things are stable.

ramonjd commented:
Follow ups here: https://github.com/WordPress/gutenberg/pull/42878

ramonjd commented:
Follow up for enqueuing over here:
- https://github.com/WordPress/gutenberg/pull/42880

matiasbenedetto commented:
:wave: This PR seems to break the vertical orientation of the row block for example. I reported the problem here: https://github.com/WordPress/gutenberg/issues/42938 and submitted a potential fix in https://github.com/WordPress/gutenberg/pull/42939
